### ISSUE TITLE
LZ184 under the hood

### ISSUE THREAD

**Issue by Ishinoshita** [gcp/leela-zero#1963]
Some training data chunks of LZ184 (current best network) have been recently released and I did my usual data mining exercise on them.

#### TLDR:

I had a look at the most recent chunks (_07xx, 3200 games) and identified 4 different types of client behaviour:
- 2 already well known:
      1 explained by next branch time mnagement pruning;
      1 unexplained, but definitively fake data (1.3% of games)
- 2 new behaviours:
      1 explained by BRII now pouring games at high rate (40% of the total of games), with '-m 30' command;
      1 unexplained (10-15% of the games).

I also noted that the fraction of no-resign games ('-r 0') seems to have been increased again (at least compared to LZ177).

#### Sample sgf:
- a "p=0.25 @move0" game:   [chunk_701_game_12.zip](https://github.com/gcp/leela-zero/files/2523027/chunk_701_game_12.zip)
- a "-m30" no-resign game:   [chunk_735_game_2.zip](https://github.com/gcp/leela-zero/files/2523028/chunk_735_game_2.zip)
- a "-m 999" no-resign game (for reference):   [chunk_730_game_22.zip](https://github.com/gcp/leela-zero/files/2523029/chunk_730_game_22.zip)

#### Discussion

The new behaviour that puzzles me (10%-15% of the games) is the one corresponding to a secondary peak in the policy maximum distribution at move 0. The difference in policy distrubtion is small and seems overall innocuous, but I cannot figure out the presence of this secondary peak but for assuming weird things, like either:
- games from different networks
- games with different parameters for Dirichlet noise
- games with different parameters for MCTS, in particular a different C_PUCT

As usual, devs probably know a simpler, less fancy explanation.

**Comment by alreadydone** [comment]
Nice work and beautiful ppt! So -m30 no-resign games are indeed shorter.
The  "p=0.25 @move0" sample game (definitely fake data) looks quite like a game played with -v 37 for the first 106 moves and -v 401 afterwards.
Do you have a sample from the 10-15% as well?

**Comment by Ishinoshita** [comment]
Here are 10 such games, with p_max=0.203125 at move 0 (the mode of the secondary peak). However, this is a probabilistic list, as the two distributions overlap and there could be 'standard' games in that list.
[Secondary peak sgf.zip](https://github.com/gcp/leela-zero/files/2523088/Secondary.peak.sgf.zip)

**Comment by Splee99** [comment]
Shorter no-resign games mean that LZ184 tends to pass earlier than before. This is also confirmed by some handicap games with Petgo3 on KGS and on my computer, where LZ passes when the winrate is about 3-5%. The weird thing is that the LZ "expected" winrate could jump to more than 10% when PASS gets the most visits.

**Comment by Ishinoshita** [comment]
@Splee99 My title is a bit misleadind. My observations apply not to LZ184 but to to the *training data* generated by LZ184. Which may influence LZ185 behaviour. This been said, LZ184 has also been milked with LZ183 data, obtained a priori in the same conditions (BRII + -m 30).

LZ185 will be the second generation under this blend of training settings. If there is indeed some behaviour change, it may become more pronounced.

**Comment by wonderingabout** [comment]
> games from different networks

 @Ishinoshita 
this was my first guess when i saw the very fast games #1957 
i thought if client is using another network, even a weaker card can compute them fast

**Comment by Ishinoshita** [comment]
@Splee99
>Shorter no-resign games mean that LZ184 tends to pass earlier than before.

Well, depends on what you call 'LZ184'. LZ184 + Dirichlet + MCTS + T=0 (a priori true in endgame with '-m 30') will pass earlier than with '-m999' as it will pick Pass on soon as Pass is the best move after the search, even if Pass probability is far from 1, which is far from guaranteed with T=1. So there will be much less 'missed opportunities' to get two consecutive Pass with T=0 than with T=1. 

By itself, the network doesn't need to be different and prefer pass move to behave so with '-m30'. And since the training data generated  in a given position are unrelated to the move actually picked on that position, I don't see that such games should 'teach' the network to pass earlier, intrinsically. So I'm surprised by the change you're reporting.

**Comment by petgo3** [comment]
In my opinion early passing has disappeared when i use precision single option...

**Comment by petgo3** [comment]
Use selfplay games half precision (im using next). This may  lead to weaker games, as far as i can tell from some petgo3 games on kgs

**Comment by Glrr** [comment]
instead of -m 999, -m 200 or -m 300 is probably enough

**Comment by Ishinoshita** [comment]
By 'rational explanation', I meant 'known explanation', of course. I don't suggest supranatural forces ;-)

**Comment by roy7** [comment]
Does anyone following this thread happen to use Colab and could check their own training data or provide some game hashes so we could confirm their are working normally?

It's possible there's some sort of Colab-specific bug or issue.

**Comment by tterava** [comment]
I can provide Google Cloud Compute data if that helps

**Comment by roy7** [comment]
@tterava If you could provide a training data file like is uploaded to the server to @Ishinoshita he could check it to see if it had any data oddities vs non-Colab data sources.

**Comment by tterava** [comment]
[games.zip](https://github.com/gcp/leela-zero/files/2539120/games.zip)

Here's all the files files uploaded by my Google instance.

**Comment by tterava** [comment]
I added the --debug option so i'll have some data by tomorrow

**Comment by tterava** [comment]
@wonderingabout 
I just have a regular instance with almost no automation.

Here are some training data files: 
[debug.zip](https://github.com/gcp/leela-zero/files/2542034/debug.zip)

**Comment by wonderingabout** [comment]
@tterava 
ok

what protocol do you use to upload the archive ?

transfer.sh ? mail ? else ?

**Comment by tterava** [comment]
I run a vncserver on the instance which has Firefox browser. It's not very responsive and has some weird issues, but it gets the job done.

**Comment by Ishinoshita** [comment]
@tterava Thanks for the data. I'm afraid I don't reach any clear conclusion @roy7 .

* Comparison of 56 self-play games generated yesterday/today on Colab (mostly LZ185-generated + a bit of LZ186-generated) with 3200 self-play games of most recently publicly available training data (LZ184, chunks 25xx).
*I don't have LZ185/LZ186 data yet, so kinda compare apples and oranges at this stage, although the center of the main, 'normal' peak doesn't seems to shift too fast, a priori).*

Policy max at move 0:
![image](https://user-images.githubusercontent.com/37498331/47939062-508c0c80-dee6-11e8-9d3b-0131cbc43fa2.png)

I think it would take to compare much more more Colab games, collected on a short period of time, with the corresponding training data, once published.

**Comment by tterava** [comment]
@Ishinoshita 
Nice work. Just let me know if you need more training data from Google Cloud. I left the --debug option on so all training data is now collected.

**Comment by Ishinoshita** [comment]
@tterava If LZ186 is not demoted too fast ;-), and if you can continue collecting games for while (if it's not too heavy for you), then I redo that analysis of your Colab data against LZ186 traing data, when published.

**Comment by tterava** [comment]
@wonderingabout 
I followed these instructions: https://medium.com/google-cloud/linux-gui-on-the-google-cloud-platform-800719ab27c5
After that is very easy to transfer files out when you can use a browser. I uploaded the files straight from the instance using vnc.

**Comment by wonderingabout** [comment]
@tterava 
big thanks, actually you dont need gui, you can transfer files easily in command line as i explained here : 
https://github.com/gcp/leela-zero/issues/1943#issuecomment-430977929

i may find interesting ways to automate a shutdown-script if i look at your link, thanks again

**Comment by tterava** [comment]
@Ishinoshita 
Here's the rest of the files: 
[debug2.zip](https://github.com/gcp/leela-zero/files/2546985/debug2.zip)

Google has started blocking the GPU instances with free accounts so unfortunately there won't be any more.

**Comment by wonderingabout** [comment]
@Ishinoshita 

this one also was generated today, with ??? duration, probably fake too : 
https://zero.sjeng.org/view/77987c471c33de15e9d69e79dd8feb7a858c3d8454d8a8b71766a2027bdb854b?viewer=wgo

**Comment by Hersmunch** [comment]
Note that ??? length games can occur when they are paused and resumed. I don’t know if that is the only case though.

**Comment by Ishinoshita** [comment]
*Just for completion sake as LZ184/186 is kind of it's ancient history now.*

I did compared LZ184 games from central distribution and secondary peak distribution. I observe differences but with no firm conclusion (possibly shorter, less random ?).

I also compared 385 games self-played by LZ186 on Colab (provided by @tterava) to 3200 other games from LZ186. No difference at all.

[LZ186 – Chunks 25xx – Policy max @.pptx](https://github.com/gcp/leela-zero/files/2621699/LZ186.Chunks.25xx.Policy.max.pptx)

**Comment by alreadydone** [comment]
A strange game produced by somone's autogtp instance on Colab (and found by himself): http://zero.sjeng.org/view/fe91117bc31bd90fa1ccaa2d9f26e168bde497a7cc5f8dd753633d9a0326ea50?viewer=wgo

**Comment by Ishinoshita** [comment]
@alreadydone Strange game indeed. Here are both sgf (original from sjeng.org + reconstructed with ```train2sgf``` after locating the game in chunk 209 of LZ192 training data:
[fe91117bc31bd90fa1.zip](https://github.com/gcp/leela-zero/files/2682601/fe91117bc31bd90fa1.zip)

Apart from early san-san invasion, also very early blunders, hard to swallow, like move 18 T7.

Although I didn't take time to replay with LZ190, policy search at move 25 hardly believable:

![image](https://user-images.githubusercontent.com/37498331/50042921-7838cd80-006b-11e9-8271-051dc222da92.png)

D3 top move, fine. T2 and A19 2nd and third moves shocking... ;-).
